File: implementations/dart/lib/src/parser.dart
import 'lexer.dart';

class Parser {
  final Lexer lexer;
  late Token t;
  Parser(this.lexer) {
    t = lexer.next();
  }

  Map<String, dynamic> program() {
    if (t.type == 'do') {
      _eat('do');
      final rootValue = _value();
      if (rootValue['type'] != 'array') _fail('unexpected token', t);
      if (t.type != 'eof') _fail('unexpected token after root array', t);
      return {'type': 'Document', 'blocks': <Map<String, dynamic>>[], 'rootValue': rootValue};
    }
    final blocks = <Map<String, dynamic>>[];
    while (t.type != 'eof') {
      blocks.add(_block());
    }
    return {'type': 'Document', 'blocks': blocks};
  }

  Map<String, dynamic> _block() {
    final name = _eat('id').value as String;
    String? arg;
    if (t.type == 'string') arg = _eat('string').value as String;
    _eat('do');
    final stmts = <Map<String, dynamic>>[];
    final keys = <List<String>>[];
    while (t.type != 'end') {
      if (t.type == 'eof') _fail('missing end', t);
      stmts.add(_stmt(keys));
    }
    _eat('end');
    return {'type': 'Block', 'name': name, 'arg': arg, 'statements': stmts};
  }

  Map<String, dynamic> _stmt(List<List<String>> keys) {
    final id = _eat('id');
    if (t.type == 'do') {
      _eat('do');
      if (t.type == '[') {
        final path = <String>[id.value as String];
        _check(keys, path, id);
        final value = _value();
        _eat('end');
        return {'type': 'property', 'key': path, 'value': value};
      }
      final stmts = <Map<String, dynamic>>[];
      final inner = <List<String>>[];
      while (t.type != 'end') {
        if (t.type == 'eof') _fail('missing end', t);
        stmts.add(_stmt(inner));
      }
      _eat('end');
      return {'type': 'Block', 'name': id.value, 'arg': null, 'statements': stmts};
    }
    if (t.type == 'string') {
      String? arg;
      if (t.type == 'string') arg = _eat('string').value as String;
      _eat('do');
      final stmts = <Map<String, dynamic>>[];
      final inner = <List<String>>[];
      while (t.type != 'end') {
        if (t.type == 'eof') _fail('missing end', t);
        stmts.add(_stmt(inner));
      }
      _eat('end');
      return {'type': 'Block', 'name': id.value, 'arg': arg, 'statements': stmts};
    }
    final path = <String>[id.value as String];
    while (t.type == '.') {
      _eat('.');
      path.add(_eat('id').value as String);
    }
    _eat('=');
    _check(keys, path, id);
    return {'type': 'property', 'key': path, 'value': _value()};
  }

  Map<String, dynamic> _value() {
    if (t.type == 'string') {
      final x = _eat('string');
      return {'type': 'string', 'value': x.value};
    }
    if (t.type == 'number') {
      final x = _eat('number');
      return {'type': 'number', 'value': x.value['value'], 'raw': x.value['raw']};
    }
    if (t.type == 'bool') {
      final x = _eat('bool');
      return {'type': 'boolean', 'value': x.value};
    }
    if (t.type == '[') {
      final start = _eat('[');
      final items = <Map<String, dynamic>>[];
      if (t.type != ']') {
        while (true) {
          items.add(_value());
          if (t.type != ',') break;
          _eat(',');
          if (t.type == ']') _fail('trailing comma in array', t);
        }
      }
      if (t.type != ']') _fail('missing ]', start);
      _eat(']');
      return {'type': 'array', 'items': items};
    }
    if (t.type == 'do') {
      _eat('do');
      final stmts = <Map<String, dynamic>>[];
      final inner = <List<String>>[];
      while (t.type != 'end') {
        if (t.type == 'eof') _fail('missing end', t);
        stmts.add(_stmt(inner));
      }
      _eat('end');
      return {'type': 'block', 'name': '', 'arg': null, 'statements': stmts};
    }
    if (t.type == 'id') _fail('invalid bare identifier value', t);
    _fail('unexpected token', t);
  }

  Token _eat(String type) {
    final x = t;
    if (x.type != type) _fail('unexpected token', x);
    t = lexer.next();
    return x;
  }

  void _check(List<List<String>> keys, List<String> path, Token at) {
    for (final k in keys) {
      if (_eq(k, path) || _pref(k, path) || _pref(path, k)) {
        _fail('duplicate or conflicting key path', at);
      }
    }
    keys.add(path);
  }

  bool _eq(List<String> a, List<String> b) {
    if (a.length != b.length) return false;
    for (var i = 0; i < a.length; i++) {
      if (a[i] != b[i]) return false;
    }
    return true;
  }

  bool _pref(List<String> a, List<String> b) {
    if (a.length >= b.length) return false;
    for (var i = 0; i < a.length; i++) {
      if (a[i] != b[i]) return false;
    }
    return true;
  }

  Never _fail(String msg, Token tok) {
    throw RclError('line ${tok.line}, column ${tok.col}: $msg');
  }
}

File: implementations/csharp/src/Parser.cs
using System;
using System.Collections.Generic;

namespace RCLImpl {
public sealed class Parser {
  readonly List<Token> T;
  int P;
  public Parser(string src){ T=Lexer.Lex(src); P=0; }
  public DocumentNode Parse(){
    var d=new DocumentNode();
    if(Is(TokenKind.DO)){
      Eat(TokenKind.DO,"unexpected token");
      d.RootValue=ParseArray();
      if(!Is(TokenKind.EOF)) throw Err("unexpected token after root array");
      return d;
    }
    while(!Is(TokenKind.EOF)) d.Blocks.Add(ParseBlock());
    return d;
  }

  BlockNode ParseBlock(){
    var n=Eat(TokenKind.IDENT,"unexpected token");
    string arg=null; if(Is(TokenKind.STRING)) arg=Eat(TokenKind.STRING,"unexpected token").Text;
    Eat(TokenKind.DO,"unexpected token");
    var b=ParseBlockBody(n.Text,arg);
    Eat(TokenKind.END,"unexpected token");
    return b;
  }

  BlockNode ParseAnonymousBlock(){
    Eat(TokenKind.DO,"unexpected token");
    var b=ParseBlockBody("",null);
    Eat(TokenKind.END,"unexpected token");
    return b;
  }

  BlockNode ParseBlockBody(string name,string arg){
    var b=new BlockNode(name,arg); var seen=new List<string>();
    while(!Is(TokenKind.END)){
      if(Is(TokenKind.EOF)) throw Err("missing end");
      var cur=Eat(TokenKind.IDENT,"unexpected token"); var nx=PeekAt(1);
      if(nx.Kind==TokenKind.DO){
        var afterDo=PeekAt(2);
        if(afterDo.Kind==TokenKind.LBR){
          var key=cur.Text; ValidateKey(key,seen,cur);
          Eat(TokenKind.DO,"unexpected token");
          var v=ParseArray();
          Eat(TokenKind.END,"unexpected token");
          b.Properties[key]=v; b.PropertyOrder.Add(key);
        } else {
          P--; var c=ParseBlock(); if(c.Argument==null) b.Blocks.Add(c); else b.NamedBlocks.Add(c);
        }
      } else if(nx.Kind==TokenKind.EQ||nx.Kind==TokenKind.DOT){
        var key=ParseKey(cur.Text); ValidateKey(key,seen,cur);
        Eat(TokenKind.EQ,"unexpected token"); var v=ParseValue();
        b.Properties[key]=v; b.PropertyOrder.Add(key);
      } else if(nx.Kind==TokenKind.STRING){
        P--; var c=ParseBlock(); if(c.Argument==null) b.Blocks.Add(c); else b.NamedBlocks.Add(c);
      } else throw Err("unexpected token");
    }
    return b;
  }

  string ParseKey(string f){
    var k=f;
    while(Is(TokenKind.DOT)){ Eat(TokenKind.DOT,"unexpected token"); k+="."+Eat(TokenKind.IDENT,"unexpected token").Text; }
    return k;
  }

  IAstNode ParseValue(){
    if(Is(TokenKind.STRING)) return new StringNode(Eat(TokenKind.STRING,"unexpected token").Text);
    if(Is(TokenKind.NUMBER)) return new NumberNode(double.Parse(Eat(TokenKind.NUMBER,"unexpected token").Text,System.Globalization.CultureInfo.InvariantCulture));
    if(Is(TokenKind.IDENT)){
      var v=Eat(TokenKind.IDENT,"unexpected token").Text;
      if(v=="true"||v=="false") return new BooleanNode(v=="true");
      throw Err("invalid bare identifier value");
    }
    if(Is(TokenKind.LBR)) return ParseArray();
    if(Is(TokenKind.DO)) return ParseAnonymousBlock();
    throw Err("unexpected token");
  }

  IAstNode ParseArray(){
    Eat(TokenKind.LBR,"unexpected token"); var a=new ArrayNode();
    if(Is(TokenKind.RBR)){ Eat(TokenKind.RBR,"unexpected token"); return a; }
    a.Elements.Add(ParseValue());
    while(Is(TokenKind.COMMA)){
      Eat(TokenKind.COMMA,"unexpected token");
      if(Is(TokenKind.RBR)) throw Err("trailing comma in array");
      a.Elements.Add(ParseValue());
    }
    if(!Is(TokenKind.RBR)) throw Err("missing ]");
    Eat(TokenKind.RBR,"unexpected token"); return a;
  }

  void ValidateKey(string key,List<string> seen,Token t0){
    foreach(var ex in seen){
      if(ex==key) throw new ParseException("duplicate key",t0.Line,t0.Col);
      if(IsPrefix(ex,key)||IsPrefix(key,ex)) throw new ParseException("key-path prefix conflict",t0.Line,t0.Col);
    }
    seen.Add(key);
  }

  bool IsPrefix(string a,string b){ return a.Length<b.Length&&b.StartsWith(a+"."); }
  bool Is(TokenKind k){ return T[P].Kind==k; }
  Token PeekAt(int offset){ var idx=P+offset-1; if(idx<0) idx=0; if(idx>=T.Count) idx=T.Count-1; return T[idx]; }
  Token Peek(){ return PeekAt(1); }
  Token Eat(TokenKind k,string m){ if(!Is(k)) throw Err(m); return T[P++]; }
  ParseException Err(string m){ var x=T[Math.Min(P,T.Count-1)]; return new ParseException(m,x.Line,x.Col); }
}
}
